File: rhinocommon/snippets/py/add-texture.py
import Rhino
import scriptcontext
import System.Guid
import System.Windows.Forms.DialogResult
import System.IO.File

def AddTexture():
    # Select object to add texture
    filter = Rhino.DocObjects.ObjectType.Surface | Rhino.DocObjects.ObjectType.PolysrfFilter | Rhino.DocObjects.ObjectType.Mesh
    rc, objref = Rhino.Input.RhinoGet.GetOneObject("Select object to add texture", False, filter)
    if rc!=Rhino.Commands.Result.Success: return rc

    rhino_object = objref.Object()
    if not rhino_object: return Rhino.Commands.Result.Failure

    # Select texture
    fd = Rhino.UI.OpenFileDialog()
    fd.Filter = "Image Files (*.bmp;*.png;*.jpg)|*.bmp;*.png;*.jpg"
    if not fd.ShowDialog():
        return Rhino.Commands.Result.Cancel

    # Verify texture
    bitmap_filename = fd.FileName
    if not System.IO.File.Exists(bitmap_filename):
        return Rhino.Commands.Result.Nothing

    # Make sure the object has it's material source set to "material_from_object"
    rhino_object.Attributes.MaterialSource = Rhino.DocObjects.ObjectMaterialSource.MaterialFromObject

    # Make sure the object has a material assigned
    material_index = rhino_object.Attributes.MaterialIndex
    if material_index<0:
        # Create a new material based on Rhino's default material
        material_index = scriptcontext.doc.Materials.Add()
        # Assign the new material (index) to the object.
        rhino_object.Attributes.MaterialIndex = material_index

    if material_index>=0:
        mat = scriptcontext.doc.Materials[material_index]
        mat.SetBumpTexture(bitmap_filename)
        mat.CommitChanges()

        #Don't forget to update the object, if necessary
        rhino_object.CommitChanges()

        scriptcontext.doc.Views.Redraw()
        return Rhino.Commands.Result.Success

    return Rhino.Commands.Result.Failure

if __name__=="__main__":
    AddTexture()

File: rhinocommon/snippets/cs/add-texture.cs

using Rhino.Commands;
using Rhino.DocObjects;
using Rhino.Render;
using System.Drawing;

partial class Examples
{
  public static Result AddTexture(Rhino.RhinoDoc doc)
  {
    // This example shows how to assign a render texture to an object. The texture is created as a child
    // of a material in the bitmap-texture child slot. The texture slot is also switched on, and a blend
    // amount of 80% is set on the texture slot. This last part is optional; the default is 100%.

    // Select object to add texture.
    const ObjectType filter = ObjectType.Surface | ObjectType.PolysrfFilter | ObjectType.Mesh;
    var rc = Rhino.Input.RhinoGet.GetOneObject("Select object to add texture", false, filter, out ObjRef objref);
    if (rc != Result.Success)
      return rc;

    var rhino_object = objref.Object();
    if (rhino_object == null)
      return Result.Failure;

    // Create a Rhino material.
    var rhino_material = new Material
    {
      Name = "Material",
      DiffuseColor = Color.White
    };

    // Create a basic Render material from the Rhino material.
    var render_material = RenderMaterial.CreateBasicMaterial(rhino_material, doc);

    // Add the basic Render material to the document.
    doc.RenderMaterials.Add(render_material);

    // We will be adding a texture to the material's bitmap-texture child slot.
    var child_slot_name = "bitmap-texture";
    var cc = RenderContent.ChangeContexts.Program;

    // Create a bitmap texture. This will also prompt for a file name.
    RenderContent.Create(doc, ContentUuids.BitmapTextureType, render_material, child_slot_name);

    // Turn on the bitmap-texture child slot of the render material and set a blend amount of 80%.
    render_material.BeginChange(cc);
    render_material.SetChildSlotOn(child_slot_name, true, cc);
    render_material.SetChildSlotAmount(child_slot_name, 80.0, cc);
    render_material.EndChange();

    // Assign the render material to the object.
    rhino_object.RenderMaterial = render_material;
    rhino_object.CommitChanges();

    // Uodate the viewports.
    doc.Views.Redraw();

    return Result.Success;
  }
}
